matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/taggedevents/TaggedEventsContent.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2021 The Matrix.org Foundation C.I.C.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.matrix.android.sdk.internal.session.room.taggedevents
+
+import com.squareup.moshi.Json
+import com.squareup.moshi.JsonClass
+
+/**
+ * Keys are event IDs, values are event information.
+ */
+typealias TaggedEvent = Map<String, TaggedEventInfo>
+
+/**
+ * Keys are tagged event names (eg. m.favourite), values are the related events.
+ */
+typealias TaggedEvents = Map<String, TaggedEvent>
+
+/**
+ * Class used to parse the content of a m.tagged_events type event.
+ * This kind of event defines the tagged events in a room.
+ *
+ * The content of this event is a tags key whose value is an object mapping the name of each tag
+ * to another object. The JSON object associated with each tag is an object where the keys are the
+ * event IDs and values give information about the events.
+ *
+ * Ref: https://github.com/matrix-org/matrix-doc/pull/2437
+ */
+@JsonClass(generateAdapter = true)
+data class TaggedEventsContent(
+        var tags: TaggedEvents = emptyMap()
+) {
+    val favouriteEvents
+        get() = tags[TAG_FAVOURITE].orEmpty()
+
+    val hiddenEvents
+        get() = tags[TAG_HIDDEN].orEmpty()
+
+    fun tagEvent(eventId: String, info: TaggedEventInfo, tag: String) {
+        val tagMap = HashMap<String, TaggedEventInfo>(tags[tag] ?: emptyMap())
+        tagMap[eventId] = info
+
+        val updatedTags = HashMap<String, Map<String, TaggedEventInfo>>(tags)
+        updatedTags[tag] = tagMap
+        tags = updatedTags
+    }
```

Comment:
I propose to rewrite to
```kotlin
    fun tagEvent(eventId: String, tag: String, info: TaggedEventInfo) {
        val tagMap = tags[tag].orEmpty().toMutableMap()
                .also { it[eventId] = info }

        tags = tags.toMutableMap().also { it[tag] = tagMap }
    }
```

Reply:
I found a better optimization :)

Reply:
TIL the `plus()` (and `minus` operator extension on `Map`). Note that it can be written using `+` (and `- `) but for code clarity I prefer what you have done. Thanks!